controllers/vsphereclusteridentity_controller.go

```diff
@@ -0,0 +1,179 @@
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	clusterutilv1 "sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	identityControlledType     = &infrav1.VSphereClusterIdentity{}
+	identityControlledTypeName = reflect.TypeOf(clusterControlledType).Elem().Name()
+	identitySecretMU           sync.Mutex
+)
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=vsphereclusteridentities,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=vsphereclusteridentities/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;patch;update
+
+func AddVsphereClusterIdentityControllerToManager(ctx *context.ControllerManagerContext, mgr manager.Manager) error {
+	var (
+		controllerNameShort = fmt.Sprintf("%s-controller", strings.ToLower(identityControlledTypeName))
+		controllerNameLong  = fmt.Sprintf("%s/%s/%s", ctx.Namespace, ctx.Name, controllerNameShort)
+	)
+
+	// Build the controller context.
+	controllerContext := &context.ControllerContext{
+		ControllerManagerContext: ctx,
+		Name:                     controllerNameShort,
+		Recorder:                 record.New(mgr.GetEventRecorderFor(controllerNameLong)),
+		Logger:                   ctx.Logger.WithName(controllerNameShort),
+	}
+
+	reconciler := clusterIdentityReconciler{ControllerContext: controllerContext}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(identityControlledType).
+		WithOptions(controller.Options{MaxConcurrentReconciles: ctx.MaxConcurrentReconciles}).
+		Complete(reconciler)
+}
+
+type clusterIdentityReconciler struct {
+	*context.ControllerContext
+}
+
+func (r clusterIdentityReconciler) Reconcile(ctx _context.Context, req reconcile.Request) (_ reconcile.Result, reterr error) {
+	// Get VSphereClusterIdentity
+	identity := &infrav1.VSphereClusterIdentity{}
+	if err := r.Client.Get(r, req.NamespacedName, identity); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Logger.V(4).Info("VSphereClusterIdentity not found, won't reconcile", "key", req.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Create the patch helper.
+	patchHelper, err := patch.NewHelper(identity, r.Client)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(
+			err,
+			"failed to init patch helper for %s %s/%s",
+			identity.GroupVersionKind(),
+			identity.Namespace,
+			identity.Name)
+	}
+
+	defer func() {
+		conditions.SetSummary(identity, conditions.WithConditions(infrav1.CredentialsAvailableCondidtion))
```

Comment:
for consistency we might want to have an `identity_context.go`, see `pkg/context`

Reply:
It crossed my mind during implementation but the logic is fully contained in the controller right now and there aren't any other resources constantly used by the identity. I think it makes more sense if/when the identity logic grows

Reply:
can you add a TODO to consider it in the future ?